File: rust/cli/src/commands/tree.rs
//! `ball tree <program>` — print a Ball program's module/import tree (issue
//! #365).
//!
//! Delegates to the self-hosted `cli_core` verb `treeReport` (compiled from
//! `dart/shared/lib/cli_core.dart` via `cargo run -p ball-cli-regen` — see
//! `rust/cli/AGENTS.md`), byte-identical to `dart/cli/lib/src/runner.dart`'s
//! `_tree`. Behind the `cli_core` Cargo feature — see `Cargo.toml`.
use std::path::Path;

use crate::error::CliError;
use crate::loader::load_engine;

#[cfg(feature = "cli_core")]
pub fn tree(path: &Path) -> Result<(), CliError> {
    let engine = load_engine(path)?;
    let report = crate::compiled_cli::treeReport(engine.program_value().clone());
    println!("{report}");
    Ok(())
}

#[cfg(not(feature = "cli_core"))]
pub fn tree(path: &Path) -> Result<(), CliError> {
    let _engine = load_engine(path)?;
    Err(CliError::Runtime(
        "`ball tree` needs the self-hosted cli-core, built in via `ball-cli`'s `cli_core` \
         Cargo feature (off by default — see rust/cli/Cargo.toml). Build with `--features \
         cli_core` after regenerating rust/cli/src/compiled_cli.rs (`cargo run -p \
         ball-cli-regen`, which itself needs `dart/self_host/cli.ball.json` — see \
         rust/cli/AGENTS.md)."
            .to_string(),
    ))
}

File: go/cli/tree.go
package cli

import (
	"flag"
	"io"

	ballrt "github.com/ball-lang/ball/go/runtime"
)

// cmdTree implements `ball tree <program.ball.json>` (issue #570): print the
// program's module/import tree.
//
// Like `info`, the report comes from the self-hosted CLI core
// (cli_core.treeReport), so it is byte-identical to the Dart CLI's `tree`. The
// Go sibling of rust/cli/src/commands/tree.rs.
func cmdTree(args []string, w io.Writer) *cliError {
	const usage = "ball tree <program.ball.json>"
	view, cerr := loadCliCoreView("tree", usage, flag.NewFlagSet("tree", flag.ContinueOnError), args)
	if cerr != nil {
		return cerr
	}
	report, cerr := cliCoreTree(view)
	if cerr != nil {
		return cerr
	}
	return printLine(w, report)
}

// loadCliCoreView parses the single program-path positional every cli-core verb
// takes and returns the program's canonical proto3-JSON view — the input the
// compiled report functions consume.
//
// Loading happens BEFORE the cli-core availability check on purpose (see
// cmdInfo's doc comment): an unreadable file or a malformed program must report
// its own I/O (3) / parse (2) failure rather than being masked by the build-tag
// message, which is exactly what rust/cli's `let _engine = load_engine(path)?;`
// does in its `#[cfg(not(feature = "cli_core"))]` arms.
func loadCliCoreView(name, usage string, fs *flag.FlagSet, args []string) (view ballrt.Value, cerr *cliError) {
	positionals, cerr := parseCommand(fs, name, usage, args)
	if cerr != nil {
		return nil, cerr
	}
	if len(positionals) != 1 {
		return nil, parseErr("%s: expected exactly one program path (usage: %s)", name, usage)
	}
	eng, cerr := loadEngine(positionals[0])
	if cerr != nil {
		return nil, cerr
	}
	return eng.ProgramValue(), nil
}
